peps/pep-0767.rst

```diff
@@ -0,0 +1,650 @@
+
+* In ``__init__``, on the instance received as the first parameter (likely, ``self``).
+* In ``__new__``, on instances of the declaring class created via a call
+  to a super-class' ``__new__`` method.
+* At declaration in the body of the class.
+
+Additionally, a type checker may choose to allow the assignment:
+
+* In ``__new__``, on instances of the declaring class, without regard
+  to the origin of the instance.
+  (This choice trades soundness, as the instance may already be initialized,
+  for the simplicity of implementation.)
+* In ``@classmethod``\ s, on instances of the declaring class created via
+  a call to the class' or super-class' ``__new__`` method.
+
+Note that a child class cannot assign to any read-only attributes of a parent class
+in any of the aforementioned contexts, unless the attribute is redeclared.
+
+.. code-block:: python
+
+    from collections import abc
+    from typing import ReadOnly
+
+
+    class Band:
+        name: str
+        songs: ReadOnly[list[str]]
+
+        def __init__(self, name: str, songs: abc.Iterable[str] | None = None) -> None:
+            self.name = name
+            self.songs = []
+
+            if songs is not None:
+                self.songs = list(songs)  # multiple assignments are fine
+
+        def clear(self) -> None:
+            # error: assignment to read-only "songs" outside initialization
+            self.songs = []
+
+
+    band = Band(name="Bôa", songs=["Duvet"])
+    band.name = "Python"           # ok: "name" is not read-only
+    band.songs = []                # error: "songs" is read-only
+    band.songs.append("Twilight")  # ok: list is mutable
+
+
+    class SubBand(Band):
+        def __init__(self) -> None:
+            self.songs = []  # error: cannot assign to a read-only attribute of a base class
+
+.. code-block:: python
+
+    # a simplified immutable Fraction class
+    class Fraction:
+        numerator: ReadOnly[int]
+        denominator: ReadOnly[int]
+
+        def __new__(
+            cls,
+            numerator: str | int | float | Decimal | Rational = 0,
+            denominator: int | Rational | None = None
+        ) -> Self:
+            self = super().__new__(cls)
+
+            if denominator is None:
+                if type(numerator) is int:
+                    self.numerator = numerator
+                    self.denominator = 1
+                    return self
+
+                elif isinstance(numerator, Rational): ...
+
+            else: ...
+
+        @classmethod
+        def from_float(cls, f: float, /) -> Self:
+            self = super().__new__(cls)
+            self.numerator, self.denominator = f.as_integer_ratio()
+            return self
+
+When a class-level declaration has an initializing value, it can serve as a `flyweight <https://en.wikipedia.org/wiki/Flyweight_pattern>`_
+default for instances:
+
+.. code-block:: python
+
+    class Patient:
+        number: ReadOnly[int] = 0
```

Comment:
Are the semantics of omitting the type explained? Example:
```py
    class Patient:
        number: ReadOnly = 0  # Is this ok? What's the type of number?
```
If type inference is supported, it would be useful to have some guidelines about how type checkers should approach this, even if some details are omitted (similar to other PEPs that don't specify type inference in detail).

Reply:
There was some discussion on this here: https://github.com/python/peps/pull/4127#discussion_r1849261608